libraries/AP_MultiHeap/AP_MultiHeap.cpp

```diff
@@ -0,0 +1,201 @@
+void MultiHeap::destroy(void)
+{
+    if (!available()) {
+        return;
+    }
+    for (uint8_t i=0; i<num_heaps; i++) {
+        if (heaps[i].hp != nullptr) {
+            heap_destroy(heaps[i].hp);
+            heaps[i].hp = nullptr;
+        }
+    }
+    delete[] heaps;
+    heaps = nullptr;
+    num_heaps = 0;
+    sum_size = 0;
+}
+
+// return true if heap is available for operations
+bool MultiHeap::available(void) const
+{
+    return heaps != nullptr && heaps[0].hp != nullptr;
+}
+
+/*
+  allocate memory from a heap
+ */
+void *MultiHeap::allocate(uint32_t size)
+{
+    if (!available() || size == 0) {
+        return nullptr;
+    }
+    for (uint8_t i=0; i<num_heaps; i++) {
+        if (heaps[i].hp == nullptr) {
+            break;
+        }
+        void *newptr = heap_allocate(heaps[i].hp, size);
+        if (newptr != nullptr) {
+            last_failed = false;
+            return newptr;
+        }
+    }
+    if (!allow_expansion || !last_failed) {
+        /*
+          we only allow expansion when the last allocation
+          failed. This gives the lua engine a chance to use garbage
+          collection to recover memory
+         */
+        last_failed = true;
+        return nullptr;
+    }
+
+    if (!hal.util->get_soft_armed()) {
+        // only expand the available heaps when armed. When disarmed
+        // user should fix their SCR_HEAP_SIZE parameter
+        last_failed = true;
+        return nullptr;
+    }
+
+    /*
+      vehicle is armed and MultiHeap (for scripting) is out of
+      memory. We will see if we can add a new heap from available
+      memory if we have at least reserve_size bytes free
+     */
+    const uint32_t available = hal.util->available_memory();
+    const uint32_t heap_overhead = 128; // conservative value, varies with HAL
+    const uint32_t min_size = size + heap_overhead;
+    if (available < reserve_size+min_size) {
+        last_failed = true;
+        return nullptr;
+    }
+
+    // round up to a minimum of 30k to allocate, and allow for heap overhead
+    const uint32_t round_to = 30*1024U;
+    const uint32_t alloc_size = MIN(available - reserve_size, MAX(size+heap_overhead, round_to));
+    if (alloc_size < min_size) {
+        last_failed = true;
+        return nullptr;
+    }
+    for (uint8_t i=0; i<num_heaps; i++) {
+        if (heaps[i].hp == nullptr) {
+            heaps[i].hp = heap_create(alloc_size);
+            if (heaps[i].hp == nullptr) {
+                last_failed = true;
+                return nullptr;
+            }
+            sum_size += alloc_size;
+            expanded_to = sum_size;
+            void *p = heap_allocate(heaps[i].hp, size);
+            last_failed = p == nullptr;
+            return p;
+        }
+    }
+    last_failed = true;
+    return nullptr;
+}
+
+/*
+  free memory from a heap
+ */
+void MultiHeap::deallocate(void *ptr)
+{
+    if (!available() || ptr == nullptr) {
+        return;
+    }
+    last_failed = false;
```

Comment:
I don't think this is valid. Suppose Lua fails an allocation, then GCs some memory, but not enough. You need to expand on the next allocation because that's the last chance you get. (I don't believe the GC will ever allocate anything)

Reply:
i think you're right, changed, thanks